src/control/src/pure_pursuit/pp.cpp

```diff
@@ -0,0 +1,144 @@
+  double min_distance = 1e9;
+  Point closest_point = Point();
+  Point aux_point = Point();
+  int closest_point_id = -1;
+  for (long unsigned int i = 0; i < path_msg->pathpoint_array.size(); i++) {
+    aux_point = Point(path_msg->pathpoint_array[i].x, path_msg->pathpoint_array[i].y);
+    double distance = this->vehicle_pose_.rear_axis_.euclidean_distance(aux_point);
+    if (distance < min_distance) {
+      min_distance = distance;
+      closest_point = aux_point;
+      closest_point_id = i;
+    }
+  }
+  return std::make_pair(closest_point, closest_point_id);
+}
+
+/**
+ * @brief Update Lookahead point
+ *
+ * @param path
+ * @return std::pair<Point, int> lookahead point and error status (1 = error)
+ */
+std::pair<Point, bool> PurePursuit::update_lookahead_point(
+    const custom_interfaces::msg::PathPointArray::ConstSharedPtr &path_msg) {
+  Point lookahead_point = Point();
+  Point aux_point = Point();
+  for (int i = this->closest_point_id_; i < path_msg->pathpoint_array.size(); i++) {
+    aux_point = Point(path_msg->pathpoint_array[i].x, path_msg->pathpoint_array[i].y);
+    double distance = this->closest_point_.euclidean_distance(aux_point);
+    if (distance <= this->ld_ * (1 - this->ld_margin_) &&
+        distance >= this->ld_ * (1 + this->ld_margin_)) {
+      lookahead_point = aux_point;
+      return std::make_pair(lookahead_point, 0);
+    }
+  }
+  for (int i = 0; i < this->closest_point_id_; i++) {
+    aux_point = Point(path_msg->pathpoint_array[i].x, path_msg->pathpoint_array[i].y);
+    double distance = this->closest_point_.euclidean_distance(aux_point);
+    if (distance > this->ld_) {
+      lookahead_point = aux_point;
+      return std::make_pair(lookahead_point, 0);
+    }
+  }
+  return std::make_pair(lookahead_point, 1);
+}
+
+double PurePursuit::pp_steering_control_law() {
+  double alpha = calculate_alpha(this->vehicle_pose_.rear_axis_, this->vehicle_pose_.cg_,
+                                 this->lookahead_point_, this->dist_cg_2_rear_axis_);
+
+  // update lookahead distance to the actual distance
+  this->ld_ = this->vehicle_pose_.rear_axis_.euclidean_distance(this->lookahead_point_);
+
+  return atan(2 * this->wheel_base_ * sin(alpha) / this->ld_);
+}
+
+double PurePursuit::calculate_alpha(Point vehicle_rear_wheel, Point vehicle_cg,
+                                    Point lookahead_point, double dist_cg_2_rear_axis) {
+  double lookhead_point_2_rear_wheel = vehicle_rear_wheel.euclidean_distance(lookahead_point);
+  double lookhead_point_2_cg = lookahead_point.euclidean_distance(vehicle_cg);
+
+  // Law of cosines
+  double alpha = acos((pow(lookhead_point_2_rear_wheel, 2) + pow(dist_cg_2_rear_axis, 2) -
+                       pow(lookhead_point_2_cg, 2)) /
+                      (2 * lookhead_point_2_rear_wheel * dist_cg_2_rear_axis));
+
+  return alpha;
+}
+
+Point PurePursuit::cg_2_rear_axis(Point cg, double heading, double dist_cg_2_rear_axis) {
```

Comment:
This could possibly be extracted to a file in common_lib I have headed in the other PR with bicycle model formulas